### PR TITLE
ref(models): `create_or_update()` -> `update_or_create()`

### PR DESCRIPTION
**Pull request by mgaeta**
This PR drops in [`update_or_create()` from django 2+](https://docs.djangoproject.com/en/4.0/ref/models/querysets/#update-or-create) as a replacement for `create_or_update()` and fixes up some behavior. My theory is that we created `create_or_update()` when we were [still on django 1](https://github.com/getsentry/sentry/commit/af97d0fb99948f6e08e8d3ac874cb9ae1b7916b8).

The old helper was doing three things that we'd lose and have to do manually:
- swallowing `IntegrityError`s that would be thrown if the created object would conflict
- separating "update kwargs" from "create kwargs"
- resolving expression values like `col_name=F("col_name") + 1`

I think these tradeoffs are worth the consistency of the standard django query API.

See https://github.com/getsentry/getsentry/pull/7134.

### CODE REVIEW

File: src/sentry/db/models/query.py [hunk 1]
```diff
@@ -49,62 +65,55 @@ def update(self: Model, using: str | None = None, **kwargs: Any) -> int:
...
+    create_kwargs.update(
+        {_handle_key(model, k, v): _handle_value(instance, v) for k, v in defaults.items()}
+    )
 
     try:
         with transaction.atomic(using=using):
             return objects.create(**create_kwargs), True
     except IntegrityError:
-        affected = objects.filter(**kwargs).update(**values)
+        pass
```
Comment (wedamija) [inline src/sentry/db/models/query.py:107]: Why swallow the `IntegrityError` here instead of performing an update?

**Review by wedamija — CHANGES_REQUESTED** [review]
There are too many unrelated changes in this pr. Please split this up into something like:

 - Adding your new `update_or_create` method and and changes to `update`
 - Changes to managers where you move things like `create_or_save` from the model to the manager
 - Formatting changes to files that were otherwise unchanged
 - Changes from `create_or_update` -> `update_or_create`

**Review by untitaker — CHANGES_REQUESTED** [review]
Agreed with Dan. Please split up this PR by type of change but additionally by component touched/reviewer required. In addition to this being a risk to deploy, I am not interested in being subscribed to gigantic PRs that have half of the company commenting on it if it can be avoided to easily.

File: src/sentry/db/models/query.py (hunk 1 above)
Comment (mgaeta) [reply src/sentry/db/models/query.py:107]: On line 92 we do:
```py
affected = objects.filter(**kwargs).update(**defaults)
if affected:
  return affected, False
```
so on line 107 we already know that this `update()` won't do anything and we fall through to 
```py
return affected, False
```

**Pull request closed without merging**